### PR TITLE
Resolves #4183 - Implemementing context-based MIME type sniffing

### LINKED ISSUE #4183: Add context-based MIME sniffing
Follow on issue to #3144

Context based mime sniffing is described in https://mimesniff.spec.whatwg.org/#context-specific-sniffing. Certain types (such as fonts) can only be sniffed from a context. However, this will require more metadata to be handed to the sniffer task.

### PR DESCRIPTION
**Pull request by Yoric**
As mentioned in #4183, I'm not sure how to test this.

[Review on Reviewable](https://reviewable.io/reviews/servo/servo/7842)

### CODE REVIEW

File: components/net/mime_classifier.rs [hunk 1]
```diff
@@ -33,36 +35,112 @@ pub enum NoSniffFlag {
...
+                    None => Some(("application".to_owned(), "octet-stream".to_owned())),
+                    _ => supplied_type.clone()
+                }
+            },
+            LoadContext::Style => {
+                // 8.5 Sniffing in a style context
+                // 1. If the supplied MIME type is undefined, [unspecified at the time of this implementation, so fallback to Browsing]
+                // 2. The computed MIME type is the supplied MIME type.
+                match *supplied_type {
+                    None => self.classify(LoadContext::Browsing, no_sniff_flag, apache_bug_flag, supplied_type, data),
```
Comment (eefriedman) [inline components/net/mime_classifier.rs]: The actual algorithm should be something like this:

1. If the supplied MIME type is undefined, the computed MIME type is "text/css". 
2. The computed MIME type is the supplied MIME type.

File: components/net/mime_classifier.rs [hunk 2]
```diff
@@ -33,36 +35,112 @@ pub enum NoSniffFlag {
...
+                    _ => supplied_type.clone()
+                }
+            },
+            LoadContext::Script => {
+                // 8.6 Sniffing in a script context
+                // 1. If the supplied MIME type is undefined, [unspecified at the time of this implementation, so fallback to Browsing]
+                // 2. The computed MIME type is the supplied MIME type.
+                match *supplied_type {
+                    None => self.classify(LoadContext::Browsing, no_sniff_flag, apache_bug_flag, supplied_type, data),
```
Comment (eefriedman) [inline components/net/mime_classifier.rs]: Script should default to "application/javascript".

**Comment by eefriedman** [comment]
In terms of testing this... we don't actually examine the mime type in non-browser contexts, so I'm pretty sure it's impossible to test without other changes.  That said, it would probably be pretty easy to add some sensitivity.  For example, if we only tried to render images which have a known mimetype, you could check that in an `` tag an image served as `text/html` doesn't get rendered, but an image served as `application/octet-stream` does.

File: components/net/mime_classifier.rs (hunk 1 above)
Comment (Yoric) [inline components/net/mime_classifier.rs]: Fine with me. I just wanted to make sure that I didn't accidentally add something incompatible.

File: components/net/mime_classifier.rs (hunk 2 above)
Comment (Yoric) [inline components/net/mime_classifier.rs]: What, no vbscript :) ?
I'll do that.

**Comment by Yoric** [comment]
> That said, it would probably be pretty easy to add some sensitivity. For example, if we only tried to render images which have a known mimetype, you could check that in an  tag an image served as text/html doesn't get rendered, but an image served as application/octet-stream does.

Do you think we want this in Servo? Perhaps if sniffing is deactivated?

File: components/net/mime_classifier.rs [hunk 3]
```diff
@@ -33,36 +35,163 @@ pub enum NoSniffFlag {
...
+                // with the resource header as the byte sequence to be
+                // matched.
+                //
+                // 3. If image-type-matched is not undefined, the
+                // computed MIME type is image-type-matched. Abort
+                // these steps.
+                //
+                // 4. The computed MIME type is the supplied MIME type.
+                match MIMEClassifier::maybe_get_media_type(supplied_type) {
+                    Some(MediaType::Xml) | Some(MediaType::Html) => supplied_type.clone(),
```
Comment (eefriedman) [inline components/net/mime_classifier.rs]: "1. If the supplied MIME type is an XML type," doesn't mention HTML.

Comment (Yoric) [inline components/net/mime_classifier.rs]: My bad, misread.

**Comment by eefriedman** [comment]
> Do you think we want this in Servo? Perhaps if sniffing is deactivated?

For testing purposes, perhaps only if sniffing is activated (which will be the default eventually).

**Comment by Yoric** [comment]
I have applied the change you suggested to the image cache, but I haven't been able to see any different behavior. I suspect that the actual culprit is actually my lack of mastery of Apache, though, as Rust and Firefox seem to have the exact same behavior.

What's the next step?

**Comment by Yoric** [comment]
Rebased. @jdm, I'm waiting for your instructions.

**Comment by jdm** [comment]
For testing, you could add a MIME type check to the font-loading code (with a server that returns no Content-Type), and look at the `Content-Type` response header for an XHR to the same font.
-S-awaiting-review +S-needs-code-changes 


---

Reviewed 12 of 12 files at r1, 3 of 3 files at r2, 5 of 5 files at r3.
Review status: all files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

<sup>**[components/net/http_loader.rs, line 141 \[r1\]](https://reviewable.io:443/reviews/servo/servo/7842#-K0tEMyaT-MRGK_Z6vx1)** ([raw file](https://github.com/servo/servo/blob/de8f67588c28b0e3ea8ceaa5db14fee616dc9a86/components/net/http_loader.rs#L141)):</sup>
The should be the same context as the original load.


---

<sup>**[components/net/mime_classifier.rs, line 76 \[r1\]](https://reviewable.io:443/reviews/servo/servo/7842#-K0tIHY1fHqTLEYNCnvg)** ([raw file](https://github.com/servo/servo/blob/de8f67588c28b0e3ea8ceaa5db14fee616dc9a86/components/net/mime_classifier.rs#L76)):</sup>
It's generally preferred to have comments like `// Step 1` rather than quoting the full spec text.


---

<sup>**[components/net/mime_classifier.rs, line 79 \[r1\]](https://reviewable.io:443/reviews/servo/servo/7842#-K0tIb7MZBZRFPMyjBe8)** ([raw file](https://github.com/servo/servo/blob/de8f67588c28b0e3ea8ceaa5db14fee616dc9a86/components/net/mime_classifier.rs#L79)):</sup>
nit: `.` should be on the next line


---

<sup>**[components/net/mime_classifier.rs, line 90 \[r1\]](https://reviewable.io:443/reviews/servo/servo/7842#-K0tIlygmrKMwKmPyiow)** ([raw file](https://github.com/servo/servo/blob/de8f67588c28b0e3ea8ceaa5db14fee616dc9a86/components/net/mime_classifier.rs#L90)):</sup>
Remove the HTML here.


---

<sup>**[components/net/mime_classifier.rs, line 109 \[r1\]](https://reviewable.io:443/reviews/servo/servo/7842#-K-jBjxkeEKAI5ynVaon-r1-109)** ([raw file](https://github.com/servo/servo/blob/de8f67588c28b0e3ea8ceaa5db14fee616dc9a86/components/net/mime_classifier.rs#L109)):</sup>
@elifriedman Do you have sources to back this up? MIME sniffing is really hairy; I'd rather stick to implementing the precise spec text to start (i.e. don't take any action here)


---

<sup>**[components/net/mime_classifier.rs, line 118 \[r1\]](https://reviewable.io:443/reviews/servo/servo/7842#-K-jBjxkeEKAI5ynVaon-r1-118)** ([raw file](https://github.com/servo/servo/blob/de8f67588c28b0e3ea8ceaa5db14fee616dc9a86/components/net/mime_classifier.rs#L118)):</sup>
This does not sound like a good idea to me; script sniffing-related errors are some of the most easily exploited. I'd rather take no action at all until the spec is clarified.


---

<sup>**[components/net/mime_classifier.rs, line 228 \[r1\]](https://reviewable.io:443/reviews/servo/servo/7842#-K0tErsS9SoMZLQ1YxT0)** ([raw file](https://github.com/servo/servo/blob/de8f67588c28b0e3ea8ceaa5db14fee616dc9a86/components/net/mime_classifier.rs#L228)):</sup>
supplied_type.map(|(ref media_type, ref media_subtype)| {
    MIMEClassifier::get_media_type(media_type, media_subtype)
})


---

<sup>**[components/net_traits/lib.rs, line 137 \[r2\]](https://reviewable.io:443/reviews/servo/servo/7842#-K0tJyG4C4pZSFc8iEzy)** ([raw file](https://github.com/servo/servo/blob/3b26540de219ab31fd252d690cf3d1b933c1108e/components/net_traits/lib.rs#L137)):</sup>
Let's use `///` so it shows up in the docs.


---

<sup>**[components/script/document_loader.rs, line 41 \[r1\]](https://reviewable.io:443/reviews/servo/servo/7842#-K0tFMLHPEWge3RfMPKy)** ([raw file](https://github.com/servo/servo/blob/de8f67588c28b0e3ea8ceaa5db14fee616dc9a86/components/script/document_loader.rs#L41)):</sup>
nit: | on the previous line, please.


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/7842)

**Comment by eefriedman** [comment]
Review status: all files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

<sup>**[components/net/mime_classifier.rs, line 109 \[r1\]](https://reviewable.io:443/reviews/servo/servo/7842#-K-jBjxkeEKAI5ynVaon-r1-109)** ([raw file](https://github.com/servo/servo/blob/de8f67588c28b0e3ea8ceaa5db14fee616dc9a86/components/net/mime_classifier.rs#L109)):</sup>
Firefox also picks `text/css` for an unspecified MIME type; see http://mxr.mozilla.org/mozilla-central/source/layout/style/Loader.cpp#925 .

Granted, this is known to be a tricky area; sniffing CSS can lead to cross-origin attacks.  See http://blogs.msdn.com/b/ie/archive/2010/10/26/mime-handling-changes-in-internet-explorer.aspx etc.

If we wanted to be extremely conservative, we could sniff to text/plain if the MIME type is unspecified.


---

<sup>**[components/net/mime_classifier.rs, line 118 \[r1\]](https://reviewable.io:443/reviews/servo/servo/7842#-K-jBjxkeEKAI5ynVaon-r1-118)** ([raw file](https://github.com/servo/servo/blob/de8f67588c28b0e3ea8ceaa5db14fee616dc9a86/components/net/mime_classifier.rs#L118)):</sup>
Err, I'm not sure exactly what you think the risk is here?  Firefox doesn't check the MIME type of JavaScript at all.


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/7842)

**Comment by jdm** [comment]
Rebased in #8190.

**Pull request closed without merging**